### PR TITLE
Import SSH keys from launchpad

### LINKED ISSUE #3142: JaaS public SSH key import list does not include Launchpad
As per https://bugs.launchpad.net/juju/+bug/1711305, in JAAS, SSH dropdown does not include launchpad import.

### PR DESCRIPTION
**Pull request by makyo**
fixes #3142

### CODE REVIEW

**Comment by makyo** [comment]
# QA

Create a model. In the deployment flow, add a launchpad username, but no other keys. After deployment, SSH into a unit from a machine with private key matching a public key on Launchpad.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/8013/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/8014/

File: jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js [hunk 1]
```diff
@@ -229,6 +230,16 @@ class DeploymentFlow extends React.Component {
   }
 
   /**
+    Store the provided Launchpad usernames in state.
+
+    @method _setLaunchpadUsernames
+    @param {Array} usernames The list of Launchpad usernames
```
Comment (frankban) [inline jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js]: Final dot.

**Comment by urosj** [comment]
+1 that username name should be used  from the macaroon. This will work in every case. So, user should be able to just chane it. Not all users are LP users, import will just fail when they are invalid (in case of pure sso user). We should also track the fail vs success ratio (comulative metric).

This addition can be dobe in new PR.

**Comment by urosj** [comment]
Potential danger of not using a name from the macaroon is that another person accidentally gets access.

**Comment by urosj** [comment]
Adding users, what if we provide link to LP so that one can check it's correct user? Same with GH? (Again ne PR)

File: jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js [hunk 2]
```diff
@@ -344,6 +355,22 @@ class DeploymentFlow extends React.Component {
         },
         ecsOptions);
     }
+    if (this.state.lpUsernames.length) {
+      this.props.importSSHKeys(
+        this.props.getUserName(),
+        this.state.lpUsernames.map(key => `lp:${key}`),
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js]: s/key/username 🤷‍♂️

File: jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js [hunk 3]
```diff
@@ -344,6 +355,22 @@ class DeploymentFlow extends React.Component {
...
+        (error, data) => {
+          if (!error) {
+            return;
+          }
+          this.props.addNotification({
+            title: 'Cannot import SSH keys',
+            message: `Cannot import SSH keys: ${error}`,
+            level: 'error'
+          });
+        }
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js]: Can you add `console.log(error)` here as well plz

**Comment by makyo** [comment]
:shipit: with a followup branch on the way.

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**